tmk_core/common/chibios/eeprom_stm32.c

```diff
@@ -48,72 +58,124 @@ uint16_t EEPROM_Init(void) {
+        EEPROM_Clear();
+        EEPROM_Restore();
     } else {
-        // Copy Page to a buffer
-        memcpy(DataBuf, (uint8_t *)FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE), FEE_PAGE_SIZE);  // !!! Calculate base address for the desired page
-
-        // check if new data is differ to current data, return if not, proceed if yes
-        if (DataByte == *(__IO uint8_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address))) {
-            return 0;
-        }
-
-        // manipulate desired data byte in temp data array if new byte is differ to the current
-        DataBuf[FEE_ADDR_OFFSET(Address) % FEE_PAGE_SIZE] = DataByte;
-
-        // Erase Page
-        FlashStatus = FLASH_ErasePage(FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE));
-
-        // Write new data (whole page) to flash if data has been changed
-        for (i = 0; i < (FEE_PAGE_SIZE / 2); i++) {
-            if ((__IO uint16_t)(0xFF00 | DataBuf[FEE_ADDR_OFFSET(i)]) != 0xFFFF) {
-                FlashStatus = FLASH_ProgramHalfWord((FEE_PAGE_BASE_ADDRESS + (page * FEE_PAGE_SIZE)) + (i * 2), (uint16_t)(0xFF00 | DataBuf[FEE_ADDR_OFFSET(i)]));
-            }
-        }
+        FLASH_ProgramHalfWord(FEE_ADDR_ADDRESS(addr), Address);
+        FLASH_ProgramHalfWord(FEE_DATA_ADDRESS(addr), DataByte);
     }
+
     return FlashStatus;
 }
 /*****************************************************************************
  *  Read once data byte from a specified address.
  *******************************************************************************/
 uint8_t EEPROM_ReadDataByte(uint16_t Address) {
-    uint8_t DataByte = 0xFF;
+    if (!IS_VALID_ADDRESS(Address)) {
+        return FEE_EMPTY_BYTE;
+    }
 
-    // Get Byte from specified address
-    DataByte = (*(__IO uint8_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address)));
+    // Get Byte from caches
+    return DataBuf[Address];
+}
 
-    return DataByte;
+/*****************************************************************************
+ *  helper functions
+ *******************************************************************************/
+// backup the current data
+void EEPROM_Backup(void)
+{
+    uint32_t begin = FEE_PAGE_BASE_ADDRESS;
+    uint32_t end = FEE_PAGE_END_ADDRESS;
+    memset(&DataBuf[0], FEE_EMPTY_BYTE, sizeof(DataBuf));
+    while( begin < end) {
+        uint16_t addr = *(__IO uint16_t*)(FEE_ADDR_ADDRESS(begin));
+        if (IS_VALID_ADDRESS(addr)) {
+            DataBuf[addr] = *(__IO uint16_t*)(FEE_DATA_ADDRESS(begin));
+        } else if( addr == FEE_EMPTY_WORD) {
+            uint16_t data = *(__IO uint16_t*)(FEE_DATA_ADDRESS(begin));
+            if (data == FEE_EMPTY_WORD) {
+                // we reached the end of valid data
+                break;
+            }
+        }
+        begin += 4;
+    }
+}
+// restore data from DataBuf
+FLASH_Status EEPROM_Restore(void) {
+    uint32_t cur = FEE_PAGE_BASE_ADDRESS;
+    for (uint8_t i = 0; i < FEE_MAX_BYTES; i++) {
```

Comment:
this should not be uint8_t as FEE_MAX_BYTES is 1024, it would loop forever and likely brick the board